File: Java/reorder-list.java
//  Reorder List

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// Time O(N^2), Space O(1)
class Solution {
    public void reorderList(ListNode head) {
        if (head==null || head.next==null)
            return;
        ListNode p = head.next, pre = head;
        while (p != null) {
            p = reverseList(p);
            pre.next = p;
            pre = p;
            p = p.next;
        }
    }
    
    private ListNode reverseList(ListNode head) {
        if (head==null || head.next==null)
            return head;
        ListNode pre = null, cur = head, nxt = null;
        while (cur != null) {
            nxt = cur.next;
            cur.next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }
}

File: leetcode/reorder-list.cc
//  Reorder List

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head || !head->next || !head->next->next)
            return;
        ListNode *cur = head;
        while (cur->next) {
            cur->next = reverseList(cur->next);
            cur = cur->next;
        }
    }
    
    ListNode *reverseList(ListNode *head) {
        if (!head || !head->next)
            return head;
        ListNode *pre = nullptr, *cur = head, *nxt;
        while (cur) {
            nxt = cur->next;
            cur->next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }
};
